tools/c7n_mailer/c7n_mailer/slack_delivery.py

```diff
@@ -0,0 +1,134 @@
+                                                                     self.logger, 'slack_template', 'slack_default')
+                self.logger.debug("Sending to owners.")
+            elif self.email_handler.target_is_email(e.split('slack://', 1)[1]):
+                resolved_addrs = self.retrieve_user_im([e.split('slack://', 1)[1]])
+                for address, slack_target in resolved_addrs.iteritems():
+                    slack_messages[address] = get_rendered_jinja(
+                        slack_target, sqs_message, to_addrs_to_resources_map.values()[0],
+                        self.logger, 'slack_template', 'slack_default')
+                self.logger.debug("Sending to specified email target, based on lookup via Slack API.")
+            elif e.startswith('slack://#'):
+                resolved_addrs = e.split('slack://#', 1)[1]
+                slack_messages[resolved_addrs] = get_rendered_jinja(resolved_addrs, sqs_message,
+                                                                    to_addrs_to_resources_map.values()[0],
+                                                                    self.logger, 'slack_template', 'slack_default')
+
+                self.logger.debug("Sending to specified Slack channel.")
+
+        return slack_messages
+
+    def slack_handler(self, sqs_message, slack_messages):
+        for key, payload in slack_messages.iteritems():
+            self.logger.info("Sending account:%s policy:%s %s:%s slack:%s to %s" % (
+                sqs_message.get('account', ''),
+                sqs_message['policy']['name'],
+                sqs_message['policy']['resource'],
+                str(len(sqs_message['resources'])),
+                sqs_message['action'].get('slack_template', 'slack_default'),
+                json.loads(payload)["channel"])
+            )
+
+            self.send_slack_msg(payload)
+
+    def retrieve_user_im(self, email_addresses):
+        list = {}
+
+        for address in email_addresses:
+            if self.cache_engine == 'redis' and self.caching.get(address):
+                    self.logger.debug('Got slack metadata from local cache for: %s' % address)
+                    list[address] = self.caching.get(address)
+                    continue
+
+            response = self.client.api_call(
+                "users.lookupByEmail", email=address)
+
+            if not response["ok"] and "Retry-After" in response["headers"]:
+                self.logger.info("Slack API rate limiting. Waiting %d seconds") % (int(response.headers['retry-after']))
+                time.sleep(int(response.headers['Retry-After']))
+                continue
+            elif not response["ok"] and response["error"] == "invalid_auth":
+                raise Exception("Invalid Slack token.")
+            elif not response["ok"] and response["error"] == "users_not_found":
+                self.logger.info("Slack user ID not found.")
+                self.caching.set(address, {})
```

Comment:
this could be a problem as well.

ideally, using a `caching` object factory would probably be better than to explicitly check against the value `redis` multiple times.